File: lib/fluxsink/vcdfluxsink.cc
#include "lib/core/globals.h"
#include "lib/config/flags.h"
#include "lib/data/fluxmap.h"
#include "lib/core/bytes.h"
#include "protocol.h"
#include "lib/fluxsink/fluxsink.h"
#include "lib/data/fluxmapreader.h"
#include "lib/fluxsink/fluxsink.pb.h"
#include "lib/config/proto.h"
#include <fstream>
#include <sys/stat.h>
#include <sys/types.h>

class VcdFluxSink : public FluxSink
{
public:
    VcdFluxSink(const VcdFluxSinkProto& config): _config(config) {}

public:
    void writeFlux(int track, int head, const Fluxmap& fluxmap) override
    {
        mkdir(_config.directory().c_str(), 0744);
        std::ofstream of(
            fmt::format(
                "{}/c{:02d}.h{:01d}.vcd", _config.directory(), track, head),
            std::ios::out | std::ios::binary);
        if (!of.is_open())
            error("cannot open output file");

        of << "$timescale 1ns $end\n"
           << "$var wire 1 i index $end\n"
           << "$var wire 1 p pulse $end\n"
           << "$upscope $end\n"
           << "$enddefinitions $end\n"
           << "$dumpvars 0i 0p $end\n";

        FluxmapReader fmr(fluxmap);
        unsigned timestamp = 0;
        unsigned lasttimestamp = 0;
        while (!fmr.eof())
        {
            unsigned ticks;
            int event;
            fmr.getNextEvent(event, ticks);
            if (fmr.eof())
                break;

            unsigned newtimestamp = timestamp + ticks;
            if (newtimestamp != lasttimestamp)
            {
                of << fmt::format("\n#{} 0i 0p\n",
                    (uint64_t)((lasttimestamp + 1) * NS_PER_TICK));
                timestamp = newtimestamp;
                of << fmt::format("#{} ", (uint64_t)(timestamp * NS_PER_TICK));
            }

            if (event & F_BIT_PULSE)
                of << "1p ";
            if (event & F_BIT_INDEX)
                of << "1i ";

            lasttimestamp = timestamp;
        }
        of << "\n";
    }

    operator std::string() const override
    {
        return fmt::format("vcd({})", _config.directory());
    }

private:
    const VcdFluxSinkProto& _config;
};

std::unique_ptr<FluxSink> FluxSink::createVcdFluxSink(
    const VcdFluxSinkProto& config)
{
    return std::unique_ptr<FluxSink>(new VcdFluxSink(config));
}

File: java/com/cowlark/fluxengine/fluxsink/VcdFluxSink.java
package com.cowlark.fluxengine.fluxsink;

import static com.cowlark.fluxengine.wiring.FluxEngine.F_BIT_INDEX;
import static com.cowlark.fluxengine.wiring.FluxEngine.F_BIT_PULSE;
import static com.cowlark.fluxengine.wiring.FluxEngine.NS_PER_TICK;

import com.cowlark.fluxengine.core.FluxEngineException;
import com.cowlark.fluxengine.data.Fluxmap;
import com.cowlark.fluxengine.data.FluxmapReader;
import com.cowlark.fluxengine.decoders.DecoderProto;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

/**
 * A flux sink which writes VCD (Value Change Dump) files, ported from
 * lib/fluxsink/vcdfluxsink.cc.
 */
public class VcdFluxSink extends FluxSink
{
    private final String directory;

    public VcdFluxSink(String directory)
    {
        this.directory = directory;
    }

    @Override
    public void addFlux(int track, int head, Fluxmap fluxmap)
    {
        try
        {
            Files.createDirectories(Path.of(directory));
        } catch (IOException e)
        {
            throw new FluxEngineException("cannot create directory '" + directory + "'");
        }

        StringBuilder sb = new StringBuilder();
        sb.append("$timescale 1ns $end\n");
        sb.append("$var wire 1 i index $end\n");
        sb.append("$var wire 1 p pulse $end\n");
        sb.append("$upscope $end\n");
        sb.append("$enddefinitions $end\n");
        sb.append("$dumpvars 0i 0p $end\n");

        FluxmapReader fmr = new FluxmapReader(fluxmap, DecoderProto.getDefaultInstance());
        long timestamp = 0;
        long lasttimestamp = 0;
        while (!fmr.eof())
        {
            FluxmapReader.Event event = fmr.getNextEvent();
            if (fmr.eof())
                break;

            long newtimestamp = timestamp + event.ticks();
            if (newtimestamp != lasttimestamp)
            {
                sb.append("\n#");
                sb.append((long) ((lasttimestamp + 1) * NS_PER_TICK));
                sb.append(" 0i 0p\n");
                timestamp = newtimestamp;
                sb.append("#");
                sb.append((long) (timestamp * NS_PER_TICK));
                sb.append(" ");
            }

            if ((event.event() & F_BIT_PULSE) != 0)
                sb.append("1p ");
            if ((event.event() & F_BIT_INDEX) != 0)
                sb.append("1i ");

            lasttimestamp = timestamp;
        }
        sb.append("\n");

        String filename = String.format("%s/c%02d.h%01d.vcd", directory, track, head);
        try
        {
            Files.write(Path.of(filename), sb.toString().getBytes());
        } catch (IOException e)
        {
            throw new FluxEngineException("cannot open output file");
        }
    }
}
